Libraries/LibDNS/Resolver.h

```diff
@@ -476,11 +629,456 @@ class Resolver {
                 lookups->remove(message.header.id);
                 return {};
             });
-            if (result.is_error()) {
+            if (result.is_error())
                 dbgln_if(DNS_DEBUG, "DNS: Received a message with no pending lookup: {}", result.error());
-                continue;
+        }
+    }
+
+    using RRSet = Vector<Messages::ResourceRecord>;
+    struct CanonicalizedRRSetWithRRSIG {
+        RRSet rrset;
+        Messages::Records::RRSIG rrsig;
+        Vector<Messages::Records::DNSKEY> dnskeys;
+    };
+
+    NonnullRefPtr<Core::Promise<bool>> validate_dnssec_chain_step(Messages::DomainName const& name, bool top_level = false)
+    {
+        dbgln_if(DNS_DEBUG, "DNS: Validating DNSSEC chain for {}", name.to_string());
+        auto promise = Core::Promise<bool>::construct();
+        // - If this is the root, we're done, just return true.
+        if (name.labels.size() == 0) {
+            promise->resolve(true);
+            return promise;
+        }
+
+        // - Lookup the SOA record for the domain.
```

Comment:
This seems quite commented. Is there a reference doc for how to properly validate a dnssec chain that could be linked here?

Reply:
IIRC not that I could find, I'll look again for something RFC-y though

Reply:
I just linked to the spec and copied the relevant bits, there's no "here's the algorithm" document.